File: deckmaster/decks/calendar/next-event.py
#!/usr/bin/env python3
"""Fetch next calendar event from GNOME EDS.

Outputs countdown or title for the deckmaster command widget.
On each call also checks alert thresholds and fires HTTP API pushes
if within 3 / 2 / 1 minute of the event.

Usage:
  next-event.py countdown   → "2h05m" (to start) or "-15m" (to end, during event)
  next-event.py title       → "Engine Team…"
  next-event.py color       → "#00ffcc;#ffffff" or "#5599ff;#ffffff" (during event)

State file: /tmp/streamdeck-next-event-alerts.json
  Tracks which thresholds have been fired per event so duplicate
  calls (countdown + title poll within same 30s window) don't double-fire.
"""
import sys
import os
import re
import json
import time
import subprocess
import urllib.request
import gi
gi.require_version('ECal', '2.0')
gi.require_version('EDataServer', '1.2')
gi.require_version('ICalGLib', '3.0')
from gi.repository import ECal, EDataServer, ICalGLib
from datetime import datetime, timezone, timedelta
from zoneinfo import ZoneInfo

# ── config ────────────────────────────────────────────────────────────────────
LOOKAHEAD_DAYS = 7
CACHE_FILE      = "/tmp/streamdeck-next-event.json"
ALERT_FILE      = "/tmp/streamdeck-next-event-alerts.json"
CACHE_TTL       = 120   # seconds before re-querying EDS
API_KEY         = 9     # stream deck key index for the calendar button
API_URL         = f"http://localhost:9990/key/{API_KEY}"

# threshold_seconds → (flash_duration, label, background_color)
ALERTS = {
    180: ("5s",  "3 MIN",  "#cc0000"),
    120: ("5s",  "2 MIN",  "#cc0000"),
    60:  ("20s", "1 MIN!", "#cc0000"),
}

# ── EDS fetch ─────────────────────────────────────────────────────────────────

# Outlook/Exchange uses Windows timezone names; map the common ones to IANA.
WINDOWS_TZ = {
    "W. Europe Standard Time":       "Europe/Berlin",
    "Central Europe Standard Time":  "Europe/Budapest",
    "Romance Standard Time":         "Europe/Paris",
    "GMT Standard Time":             "Europe/London",
    "Eastern Standard Time":         "America/New_York",
    "Central Standard Time":         "America/Chicago",
    "Mountain Standard Time":        "America/Denver",
    "Pacific Standard Time":         "America/Los_Angeles",
    "UTC":                           "UTC",
}

def ical_time_to_dt(tt, tzid=None):
    """Convert ICalGLib.Time to a UTC-aware datetime.

    libical does not recognise Windows-style TZID values and silently treats
    them as UTC, causing events to appear hours late.  When a tzid is supplied
    we resolve it through WINDOWS_TZ (or pass it straight to zoneinfo for IANA
    names) so the local wall-clock time is interpreted correctly.
    """
    naive = datetime(
        tt.get_year(), tt.get_month(), tt.get_day(),
        tt.get_hour(), tt.get_minute(), tt.get_second(),
    )
    if tzid:
        iana = WINDOWS_TZ.get(tzid, tzid)
        try:
            return naive.replace(tzinfo=ZoneInfo(iana)).astimezone(timezone.utc)
        except Exception:
            pass  # unknown TZ name — fall through to UTC
    if tt.is_utc():
        return naive.replace(tzinfo=timezone.utc)
    # Floating time: assume system local timezone
    return naive.astimezone().astimezone(timezone.utc)

def fetch_next_event():
    registry = EDataServer.SourceRegistry.new_sync(None)
    sources  = registry.list_sources(EDataServer.SOURCE_EXTENSION_CALENDAR)
    now      = datetime.now(timezone.utc)
    end      = now + timedelta(days=LOOKAHEAD_DAYS)
    now_iso  = now.strftime("%Y%m%dT%H%M%SZ")
    end_iso  = end.strftime("%Y%m%dT%H%M%SZ")
    query    = f'(occur-in-time-range? (make-time "{now_iso}") (make-time "{end_iso}"))'
    events   = []

    for source in sources:
        try:
            client = ECal.Client.connect_sync(
                source, ECal.ClientSourceType.EVENTS, 10, None)
            ok, comps = client.get_object_list_as_comps_sync(query, None)
            if not ok:
                continue
            for comp in comps:
                # Read the raw TZID from the base DTSTART *before* EDS normalises
                # it.  Outlook uses Windows-style names like "W. Europe Standard
                # Time" that libical can't resolve, so it silently treats them as
                # UTC.  We stash the original tzid in a closure so instance_cb
                # can do the conversion correctly.
                ical = comp.get_icalcomponent()
                prop = ical.get_first_property(ICalGLib.PropertyKind.DTSTART_PROPERTY)
                raw_tzid = None
                if prop:
                    param = prop.get_first_parameter(ICalGLib.ParameterKind.TZID_PARAMETER)
                    raw_tzid = param.get_tzid() if param else None

                # Also get raw TZID for DTEND
                end_prop = ical.get_first_property(ICalGLib.PropertyKind.DTEND_PROPERTY)
                raw_end_tzid = None
                if end_prop:
                    end_param = end_prop.get_first_parameter(ICalGLib.ParameterKind.TZID_PARAMETER)
                    raw_end_tzid = end_param.get_tzid() if end_param else None

                def instance_cb(ical_comp, instance_start, instance_end, _cancel, _ud,
                                tzid=raw_tzid, end_tzid=raw_end_tzid):
                    is_allday = instance_start.is_date()
                    dt_start = ical_time_to_dt(instance_start, tzid=tzid)
                    dt_end   = ical_time_to_dt(instance_end, tzid=end_tzid)
                    summary  = ical_comp.get_summary() or "?"
                    location = ical_comp.get_location() or ""
                    if dt_start <= now < dt_end:
                        # Skip all-day events as in-progress — treat them as future
                        if not is_allday:
                            events.append((dt_start.isoformat(), summary, dt_end.isoformat(), True, location))
                    elif dt_start > now:
                        # Future event (all-day events included here)
                        events.append((dt_start.isoformat(), summary, dt_end.isoformat(), False, location))
                    return True

                client.generate_instances_for_object_sync(
                    ical,
                    int(now.timestamp()), int(end.timestamp()),
                    None, instance_cb, None)
        except Exception:
            continue

    if not events:
        return None
    # Prioritise in-progress events, then sort by start time
    in_progress = [e for e in events if e[3]]
    future      = [e for e in events if not e[3]]
    in_progress.sort(key=lambda x: x[0])
    future.sort(key=lambda x: x[0])
    pick = in_progress[0] if in_progress else future[0]
    return {
        "dt": pick[0],
        "summary": pick[1],
        "end_dt": pick[2],
        "in_progress": pick[3],
        "location": pick[4] if len(pick) > 4 else "",
        "fetched": time.time(),
    }

def get_cached_event():
    try:
        if os.path.exists(CACHE_FILE):
            with open(CACHE_FILE) as f:
                data = json.load(f)
            if time.time() - data.get("fetched", 0) < CACHE_TTL:
                return data
    except Exception:
        pass
    data = fetch_next_event()
    if data:
        try:
            with open(CACHE_FILE, "w") as f:
                json.dump(data, f)
        except Exception:
            pass
    return data

# ── alerts ────────────────────────────────────────────────────────────────────

def load_alert_state():
    try:
        with open(ALERT_FILE) as f:
            return json.load(f)
    except Exception:
        return {}

def save_alert_state(state):
    try:
        with open(ALERT_FILE, "w") as f:
            json.dump(state, f)
    except Exception:
        pass

def push_alert(label, duration, bg):
    payload = json.dumps({
        "label":      label,
        "color":      "#ffffff",
        "background": bg,
        "fontsize":   18,
        "duration":   duration,
    }).encode()
    try:
        req = urllib.request.Request(
            API_URL, data=payload,
            headers={"Content-Type": "application/json"}, method="POST")
        urllib.request.urlopen(req, timeout=2)
    except Exception:
        pass

def check_alerts(event_dt_iso, remaining_secs):
    state = load_alert_state()

    # Reset fired list when the event changes
    if state.get("dt") != event_dt_iso:
        state = {"dt": event_dt_iso, "fired": []}

    fired = state.get("fired", [])
    changed = False

    for threshold, (duration, label, bg) in ALERTS.items():
        if threshold not in fired and 0 < remaining_secs <= threshold:
            push_alert(label, duration, bg)
            fired.append(threshold)
            changed = True

    if changed:
        state["fired"] = fired
        save_alert_state(state)

# ── formatting ────────────────────────────────────────────────────────────────

def format_countdown(delta):
    total_min = int(delta.total_seconds() // 60)
    if total_min < 1:
        return "now"
    h, m = divmod(total_min, 60)
    if h >= 24:
        return f"{h // 24}d"
    if h > 0:
        return f"{h}h"
    return f"{m}m"

def truncate(text, max_len=12):
    return text if len(text) <= max_len else text[:max_len - 1] + "…"

# ── main ──────────────────────────────────────────────────────────────────────

def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "countdown"
    data = get_cached_event()

    if not data:
        print("-")
        return

    dt          = datetime.fromisoformat(data["dt"])
    end_dt      = datetime.fromisoformat(data["end_dt"]) if data.get("end_dt") else None
    in_progress = data.get("in_progress", False)
    now         = datetime.now(timezone.utc)
    delta       = dt - now
    remaining   = delta.total_seconds()

    # Always check alerts on every call (deduped via state file)
    if not in_progress:
        check_alerts(data["dt"], remaining)

    if mode == "countdown":
        if in_progress and end_dt:
            # Count down to end of meeting
            end_delta = end_dt - now
            print(format_countdown(end_delta))
        else:
            print(format_countdown(delta))
    elif mode == "title":
        print(truncate(data["summary"]))
    elif mode == "color":
        if in_progress:
            print("#5599ff;#ffffff")  # blue countdown during event
        else:
            print("#ffffff;#ffffff")  # white when counting down to future event
    elif mode == "open":
        loc = data.get("location", "")
        m = re.search(r'https?://\S+', loc)
        if m:
            subprocess.Popen(["xdg-open", m.group(0)],
                             env={**os.environ, "DISPLAY": os.environ.get("DISPLAY", ":1")})

if __name__ == "__main__":
    main()

File: deckmaster/decks/calendar/next-event.js
#!/usr/bin/env bun
/**
 * next-event.js — Stream Deck main-deck calendar widget
 *
 * Reads from the fetcher daemon's JSON, renders a single SVG→PNG icon
 * showing the next calendar event with colour-coded countdown badge.
 *
 * Usage:
 *   bun next-event.js icon   → render icon, print PNG path
 *   bun next-event.js open   → open meeting URL in browser
 *
 * Events: ~/.local/share/deckblaster/calendar-day-events.json (from fetcher)
 * Image:  /tmp/streamdeck-calendar/calendar.png
 * Alerts: /tmp/streamdeck-next-event-alerts.json
 */

import sharp from "sharp";
import { mkdir, writeFile, readFile } from "fs/promises";
import { existsSync } from "fs";
import { spawn } from "child_process";
import { join, dirname } from "path";
import { homedir } from "os";

// ── config ────────────────────────────────────────────────────────────────────

const EVENTS_FILE = join(homedir(), ".local", "share", "deckblaster", "calendar-day-events.json");
const IMG_DIR     = "/tmp/streamdeck-calendar";
const IMG_PATH    = join(IMG_DIR, "calendar.png");
const ALERT_FILE  = "/tmp/streamdeck-next-event-alerts.json";
const API_KEY     = 9;
const API_URL     = `http://localhost:9990/key/${API_KEY}`;
const DECK_ROOT   = join(dirname(import.meta.path), "..", "..");

// ── design tokens ─────────────────────────────────────────────────────────────

const W = 72, H = 72, PAD = 2;
const FONT = "DejaVu Sans,sans-serif";
const BG           = "#000000";
const COLOR_NOW    = "#f85149";
const COLOR_20MIN  = "#f0883e";
const COLOR_1HR    = "#d29922";
const COLOR_4HR    = "#58a6ff";
const COLOR_FAR    = "#3fb950";
const COLOR_PAST   = "#484f58";
const TEXT_PRIMARY = "#e6edf3";
const TEXT_SECONDARY = "#8b949e";

// alert thresholds: seconds → { duration, label, bg }
const ALERTS = {
  180: { duration: "5s", label: "3 MIN", bg: "#cc0000" },
  120: { duration: "5s", label: "2 MIN", bg: "#cc0000" },
  60:  { duration: "5s", label: "1 MIN", bg: "#cc0000" },
};

// ── helpers ───────────────────────────────────────────────────────────────────

function esc(s) {
  return String(s).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function formatCountdown(ms) {
  const min = Math.floor(ms / 60000);
  if (min < 1) return "NOW";
  const h = Math.floor(min / 60);
  if (h >= 24) return `${Math.floor(h / 24)}D`;
  if (h > 0) return `${h}H`;
  return `${min}MIN`;
}

function formatTime(iso) {
  return new Date(iso).toLocaleTimeString("de-DE", { hour: "2-digit", minute: "2-digit", hour12: false });
}

function wrapText(text, maxChars = 10, maxLines = 3) {
  const words = text.split(/\s+/);
  const lines = [];
  let cur = "";
  for (const w of words) {
    if (cur && (cur + " " + w).length > maxChars) {
      lines.push(cur); cur = w;
    } else { cur = cur ? cur + " " + w : w; }
    if (lines.length >= maxLines) break;
  }
  if (cur && lines.length < maxLines) lines.push(cur);
  if (!lines.length) lines.push(text.slice(0, maxChars));
  const last = lines[lines.length - 1];
  if (last.length > maxChars) lines[lines.length - 1] = last.slice(0, maxChars - 1) + "…";
  return lines;
}

function getBarColor(evt) {
  const now = Date.now();
  const start = new Date(evt.dt).getTime();
  const end = new Date(evt.end_dt).getTime();
  if (end <= now) return COLOR_PAST;
  if (start <= now && now < end) return COLOR_NOW;
  const min = (start - now) / 60000;
  if (min <= 20) return COLOR_20MIN;
  if (min <= 60) return COLOR_1HR;
  if (min <= 240) return COLOR_4HR;
  return COLOR_FAR;
}

// ── find next event from fetcher JSON ─────────────────────────────────────────

async function getNextEvent() {
  let fdata;
  try { fdata = JSON.parse(await readFile(EVENTS_FILE, "utf8")); }
  catch { return null; }

  const now = Date.now();
  for (let offset = 0; offset < 7; offset++) {
    const events = fdata.days?.[String(offset)] ?? [];
    for (const e of events) {
      const start = new Date(e.dt).getTime();
      const end = new Date(e.end_dt).getTime();
      if (start <= now && now < end) {
        return { ...e, in_progress: true };
      }
      if (start > now) {
        return { ...e, in_progress: false };
      }
    }
  }
  return null;
}

// ── SVG ───────────────────────────────────────────────────────────────────────

function renderSvg(evt) {
  if (!evt) {
    return `<svg width="${W}" height="${H}" xmlns="http://www.w3.org/2000/svg">
  <rect width="${W}" height="${H}" fill="${BG}"/>
  <text x="${W / 2}" y="${H / 2 + 2}" font-family="${FONT}" font-size="10" fill="${TEXT_SECONDARY}" text-anchor="middle">no events</text>
</svg>`;
  }

  const now = Date.now();
  const start = new Date(evt.dt).getTime();
  const end = new Date(evt.end_dt).getTime();
  const bc = getBarColor(evt);

  let countdownText;
  if (evt.in_progress) {
    const remaining = end - now;
    countdownText = remaining < 60000 ? "NOW" : formatCountdown(remaining);
  } else {
    const delta = start - now;
    countdownText = delta < 60000 ? "NOW" : `IN ${formatCountdown(delta)}`;
  }

  const timeRange = `${formatTime(evt.dt)} – ${formatTime(evt.end_dt)}`;
  const lines = wrapText(evt.summary, 10, 3);

  const bx = PAD + 3, by = PAD + 3, bw = W - PAD * 2 - 6, bh = 14, br = 4;
  const tx = PAD + 5;

  const titleSvg = lines.map((l, i) =>
    `<text x="${tx}" y="${by + bh + 23 + i * 11}" font-family="${FONT}" font-size="9" fill="${TEXT_PRIMARY}">${esc(l)}</text>`
  ).join("\n  ");

  return `<svg width="${W}" height="${H}" xmlns="http://www.w3.org/2000/svg">
  <rect width="${W}" height="${H}" fill="${BG}"/>
  <rect x="${bx}" y="${by}" width="${bw}" height="${bh}" rx="${br}" fill="${bc}"/>
  <text x="${bx + bw / 2}" y="${by + bh - 4}" font-family="${FONT}" font-weight="bold" font-size="9" fill="#000000" text-anchor="middle">${esc(countdownText)}</text>
  <text x="${tx}" y="${by + bh + 11}" font-family="${FONT}" font-size="9" fill="${TEXT_SECONDARY}">${esc(timeRange)}</text>
  ${titleSvg}
</svg>`;
}

// ── alerts ────────────────────────────────────────────────────────────────────

async function loadAlertState() {
  try { return JSON.parse(await readFile(ALERT_FILE, "utf8")); }
  catch { return {}; }
}

async function saveAlertState(state) {
  await writeFile(ALERT_FILE, JSON.stringify(state));
}

async function pushAlert(label, duration, bg) {
  try {
    await fetch(API_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ label, background: bg, fontsize: 18, duration }),
    });
  } catch {}
}

async function checkAlerts(evt) {
  if (!evt || evt.in_progress) return;
  const remaining = (new Date(evt.dt).getTime() - Date.now()) / 1000;
  if (remaining < 0 || remaining > 200) return;

  const state = await loadAlertState();
  const dtKey = evt.dt;
  if (state.dt !== dtKey) { state.dt = dtKey; state.fired = []; }
  const fired = state.fired || [];
  let changed = false;

  for (const [thresh, { duration, label, bg }] of Object.entries(ALERTS)) {
    const t = Number(thresh);
    if (!fired.includes(t) && remaining > 0 && remaining <= t) {
      await pushAlert(label, duration, bg);
      fired.push(t);
      changed = true;
    }
  }
  if (changed) { state.fired = fired; await saveAlertState(state); }
}

// ── main ──────────────────────────────────────────────────────────────────────

const [,, mode] = process.argv;

if (mode === "open") {
  const evt = await getNextEvent();
  if (evt) {
    const m = (evt.location || "").match(/https?:\/\/\S+/);
    if (m) {
      spawn("xdg-open", [m[0]], {
        detached: true, stdio: "ignore",
        env: { ...process.env, DISPLAY: process.env.DISPLAY || ":1" },
      }).unref();
    }
  }
} else {
  // icon mode (default)
  const evt = await getNextEvent();
  await checkAlerts(evt);

  const svg = renderSvg(evt);
  await mkdir(IMG_DIR, { recursive: true });
  await sharp(Buffer.from(svg)).png().toFile(IMG_PATH);
  console.log(IMG_PATH);
}
